File: scripts/scenarios/core/agent-metareasoning-manager/api/circuit_breaker.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// CircuitBreakerState represents the state of a circuit breaker
type CircuitBreakerState int

const (
	Closed CircuitBreakerState = iota
	Open
	HalfOpen
)

func (s CircuitBreakerState) String() string {
	switch s {
	case Closed:
		return "CLOSED"
	case Open:
		return "OPEN"
	case HalfOpen:
		return "HALF_OPEN"
	default:
		return "UNKNOWN"
	}
}

// CircuitBreakerConfig holds configuration for circuit breaker
type CircuitBreakerConfig struct {
	MaxFailures     int           // Number of failures before opening
	ResetTimeout    time.Duration // Time before trying again
	SuccessThreshold int          // Successes needed in half-open to close
}

// DefaultCircuitBreakerConfig returns sensible defaults
func DefaultCircuitBreakerConfig() CircuitBreakerConfig {
	return CircuitBreakerConfig{
		MaxFailures:     3,                // Open after 3 failures
		ResetTimeout:    30 * time.Second, // Try again after 30 seconds
		SuccessThreshold: 2,               // Need 2 successes to close
	}
}

// CircuitBreaker implements the circuit breaker pattern
type CircuitBreaker struct {
	config          CircuitBreakerConfig
	state           CircuitBreakerState
	failureCount    int
	successCount    int
	lastFailureTime time.Time
	mutex           sync.RWMutex
	name            string // For logging
}

// NewCircuitBreaker creates a new circuit breaker
func NewCircuitBreaker(name string, config CircuitBreakerConfig) *CircuitBreaker {
	return &CircuitBreaker{
		config: config,
		state:  Closed,
		name:   name,
	}
}

// Call executes a function with circuit breaker protection
func (cb *CircuitBreaker) Call(fn func() error) error {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	// Check if we should allow the call
	if !cb.shouldAllowCall() {
		return fmt.Errorf("circuit breaker %s is OPEN - service unavailable", cb.name)
	}

	// Execute the function
	err := fn()
	
	// Record the result
	if err != nil {
		cb.recordFailure()
		return fmt.Errorf("circuit breaker %s recorded failure: %w", cb.name, err)
	} else {
		cb.recordSuccess()
		return nil
	}
}

// shouldAllowCall determines if a call should be allowed based on circuit breaker state
func (cb *CircuitBreaker) shouldAllowCall() bool {
	switch cb.state {
	case Closed:
		return true
	case Open:
		// Check if enough time has passed to try again
		if time.Since(cb.lastFailureTime) > cb.config.ResetTimeout {
			cb.state = HalfOpen
			cb.successCount = 0
			return true
		}
		return false
	case HalfOpen:
		return true
	default:
		return false
	}
}

// recordFailure records a failure and potentially opens the circuit
func (cb *CircuitBreaker) recordFailure() {
	cb.failureCount++
	cb.lastFailureTime = time.Now()
	
	switch cb.state {
	case Closed:
		if cb.failureCount >= cb.config.MaxFailures {
			cb.state = Open
			fmt.Printf("Circuit breaker %s OPENED after %d failures\n", cb.name, cb.failureCount)
		}
	case HalfOpen:
		cb.state = Open
		fmt.Printf("Circuit breaker %s returned to OPEN from half-open\n", cb.name)
	}
}

// recordSuccess records a success and potentially closes the circuit
func (cb *CircuitBreaker) recordSuccess() {
	cb.failureCount = 0 // Reset failure count on success
	
	switch cb.state {
	case HalfOpen:
		cb.successCount++
		if cb.successCount >= cb.config.SuccessThreshold {
			cb.state = Closed
			fmt.Printf("Circuit breaker %s CLOSED after %d successes\n", cb.name, cb.successCount)
		}
	}
}

// GetState returns the current state of the circuit breaker
func (cb *CircuitBreaker) GetState() CircuitBreakerState {
	cb.mutex.RLock()
	defer cb.mutex.RUnlock()
	return cb.state
}

// GetStats returns current statistics
func (cb *CircuitBreaker) GetStats() map[string]interface{} {
	cb.mutex.RLock()
	defer cb.mutex.RUnlock()
	
	return map[string]interface{}{
		"name":          cb.name,
		"state":         cb.state.String(),
		"failure_count": cb.failureCount,
		"success_count": cb.successCount,
		"last_failure":  cb.lastFailureTime,
	}
}

// CircuitBreakerManager manages multiple circuit breakers
type CircuitBreakerManager struct {
	breakers map[string]*CircuitBreaker
	mutex    sync.RWMutex
}

// NewCircuitBreakerManager creates a new manager
func NewCircuitBreakerManager() *CircuitBreakerManager {
	return &CircuitBreakerManager{
		breakers: make(map[string]*CircuitBreaker),
	}
}

// GetBreaker gets or creates a circuit breaker for a service
func (m *CircuitBreakerManager) GetBreaker(name string) *CircuitBreaker {
	m.mutex.RLock()
	if breaker, exists := m.breakers[name]; exists {
		m.mutex.RUnlock()
		return breaker
	}
	m.mutex.RUnlock()

	// Create new breaker
	m.mutex.Lock()
	defer m.mutex.Unlock()
	
	// Double-check after acquiring write lock
	if breaker, exists := m.breakers[name]; exists {
		return breaker
	}
	
	config := DefaultCircuitBreakerConfig()
	// Customize config based on service
	switch name {
	case "ollama":
		config.ResetTimeout = 60 * time.Second // AI services may need longer recovery
	case "n8n", "windmill":
		config.ResetTimeout = 30 * time.Second
	}
	
	breaker := NewCircuitBreaker(name, config)
	m.breakers[name] = breaker
	return breaker
}

// GetAllStats returns stats for all circuit breakers
func (m *CircuitBreakerManager) GetAllStats() map[string]interface{} {
	m.mutex.RLock()
	defer m.mutex.RUnlock()
	
	stats := make(map[string]interface{})
	for name, breaker := range m.breakers {
		stats[name] = breaker.GetStats()
	}
	
	return stats
}

// Global circuit breaker manager
var circuitBreakerManager = NewCircuitBreakerManager()

File: scenarios/browser-automation-studio/playwright-driver/src/infra/circuit-breaker.ts
/**
 * Circuit Breaker Pattern Implementation
 *
 * Prevents cascading failures by tracking consecutive failures and
 * temporarily disabling operations when a threshold is reached.
 *
 * State machine:
 * - CLOSED: Normal operation, requests pass through
 * - OPEN: Failures exceeded threshold, requests blocked
 * - HALF_OPEN: Testing if service recovered, single request allowed
 *
 * CLOSED --[failures >= threshold]--> OPEN
 * OPEN --[reset timeout elapsed]--> HALF_OPEN
 * HALF_OPEN --[success]--> CLOSED
 * HALF_OPEN --[failure]--> OPEN
 *
 * @module infra/circuit-breaker
 */

import { logger, scopedLog, LogContext, metrics } from '../utils';

// =============================================================================
// Types
// =============================================================================

/**
 * Circuit breaker configuration.
 */
export interface CircuitBreakerConfig {
  /** Number of consecutive failures before opening circuit */
  maxFailures: number;
  /** Time in ms before attempting half-open state */
  resetTimeoutMs: number;
  /** Optional name for logging */
  name?: string;
}

/**
 * Internal state for a single circuit.
 */
export interface CircuitState {
  consecutiveFailures: number;
  isOpen: boolean;
  lastFailureTime: number | null;
  totalFailures: number;
  totalSuccesses: number;
  /** Prevents multiple concurrent half-open attempts */
  halfOpenInProgress: boolean;
}

/**
 * Statistics for a circuit.
 */
export interface CircuitStats {
  isOpen: boolean;
  consecutiveFailures: number;
  totalFailures: number;
  totalSuccesses: number;
  halfOpenInProgress: boolean;
  lastFailureTime: number | null;
}

/**
 * Circuit breaker interface.
 */
export interface CircuitBreaker<K = string> {
  /** Record a successful operation - resets consecutive failure count */
  recordSuccess(key: K): void;
  /** Record a failed operation - may trip the circuit. Returns true if circuit is now open */
  recordFailure(key: K): boolean;
  /** Check if circuit is open (should skip operations) */
  isOpen(key: K): boolean;
  /** Attempt to enter half-open state. Returns true if this caller should attempt the test */
  tryEnterHalfOpen(key: K): boolean;
  /** Clean up circuit state for a key */
  cleanup(key: K): void;
  /** Get statistics for a circuit */
  getStats(key: K): CircuitStats | null;
  /** Get all active circuit keys */
  getActiveKeys(): K[];
}

// =============================================================================
// Default Configuration
// =============================================================================

export const DEFAULT_CIRCUIT_BREAKER_CONFIG: CircuitBreakerConfig = {
  maxFailures: 5,
  resetTimeoutMs: 30_000, // 30 seconds
  name: 'circuit-breaker',
};

// =============================================================================
// Implementation
// =============================================================================

/**
 * Create a new circuit breaker instance.
 *
 * Each instance manages multiple circuits keyed by a generic key type.
 * This allows a single circuit breaker to track multiple endpoints/services.
 *
 * @example
 * ```typescript
 * // Per-session callback circuit breaker
 * const callbackBreaker = createCircuitBreaker<string>({
 *   maxFailures: 5,
 *   resetTimeoutMs: 30_000,
 *   name: 'callback',
 * });
 *
 * // Usage
 * if (callbackBreaker.isOpen(sessionId) && !callbackBreaker.tryEnterHalfOpen(sessionId)) {
 *   return; // Skip callback, circuit is open
 * }
 *
 * try {
 *   await sendCallback();
 *   callbackBreaker.recordSuccess(sessionId);
 * } catch (err) {
 *   callbackBreaker.recordFailure(sessionId);
 * }
 * ```
 */
export function createCircuitBreaker<K = string>(
  config: Partial<CircuitBreakerConfig> = {}
): CircuitBreaker<K> {
  const cfg: CircuitBreakerConfig = { ...DEFAULT_CIRCUIT_BREAKER_CONFIG, ...config };
  const circuits = new Map<K, CircuitState>();

  /**
   * Get or create circuit state for a key.
   */
  function getCircuit(key: K): CircuitState {
    let state = circuits.get(key);
    if (!state) {
      state = {
        consecutiveFailures: 0,
        isOpen: false,
        lastFailureTime: null,
        totalFailures: 0,
        totalSuccesses: 0,
        halfOpenInProgress: false,
      };
      circuits.set(key, state);
    }
    return state;
  }

  return {
    recordSuccess(key: K): void {
      const state = getCircuit(key);
      state.consecutiveFailures = 0;
      state.isOpen = false;
      state.halfOpenInProgress = false;
      state.totalSuccesses++;
    },

    recordFailure(key: K): boolean {
      const state = getCircuit(key);
      state.consecutiveFailures++;
      state.totalFailures++;
      state.lastFailureTime = Date.now();

      // If we were in half-open state and failed, re-open the circuit
      if (state.halfOpenInProgress) {
        state.halfOpenInProgress = false;
        state.isOpen = true;
        logger.info(scopedLog(LogContext.RECORDING, `${cfg.name}: half-open failed, re-opening`), {
          key: String(key),
          totalFailures: state.totalFailures,
        });
        return true;
      }

      // Check if we should open the circuit
      if (state.consecutiveFailures >= cfg.maxFailures && !state.isOpen) {
        state.isOpen = true;
        logger.warn(scopedLog(LogContext.RECORDING, `${cfg.name}: circuit opened`), {
          key: String(key),
          consecutiveFailures: state.consecutiveFailures,
          totalFailures: state.totalFailures,
          hint: `Operations disabled after ${cfg.maxFailures} consecutive failures. Will retry in ${cfg.resetTimeoutMs / 1000}s.`,
        });
        // Track circuit breaker state change in metrics
        metrics.circuitBreakerStateChanges.inc({ session_id: String(key), state: 'opened' });
      }

      return state.isOpen;
    },

    isOpen(key: K): boolean {
      const state = circuits.get(key);
      return state?.isOpen ?? false;
    },

    tryEnterHalfOpen(key: K): boolean {
      const state = getCircuit(key);

      // Not open, no need for half-open
      if (!state.isOpen) {
        return false;
      }

      // Already attempting half-open
      if (state.halfOpenInProgress) {
        return false;
      }

      // Check if enough time has passed
      if (!state.lastFailureTime) {
        return false;
      }

      const elapsed = Date.now() - state.lastFailureTime;
      if (elapsed < cfg.resetTimeoutMs) {
        return false;
      }

      // Atomically claim the half-open attempt
      state.halfOpenInProgress = true;
      logger.info(scopedLog(LogContext.RECORDING, `${cfg.name}: entering half-open state`), {
        key: String(key),
        lastFailureAge: elapsed,
      });

      return true;
    },

    cleanup(key: K): void {
      circuits.delete(key);
    },

    getStats(key: K): CircuitStats | null {
      const state = circuits.get(key);
      if (!state) return null;

      return {
        isOpen: state.isOpen,
        consecutiveFailures: state.consecutiveFailures,
        totalFailures: state.totalFailures,
        totalSuccesses: state.totalSuccesses,
        halfOpenInProgress: state.halfOpenInProgress,
        lastFailureTime: state.lastFailureTime,
      };
    },

    getActiveKeys(): K[] {
      return Array.from(circuits.keys());
    },
  };
}
